app/src/main/java/fr/free/nrw/commons/nearby/fragments/NearbyParentFragment.java

```diff
@@ -1236,6 +1238,10 @@ private void setMapMarkerActions(Marker selectedMarker) {
     public void recenterMap(fr.free.nrw.commons.location.LatLng curLatLng) {
         if (curLatLng == null) {
             return;
+        } else if (!locationManager.isLocationEnabled()) {
+            // Location is not set to High Accuracy.
+            // Let's prompt the user using a dialog box.
+            showEnableLocationDialog();
         }
```

Comment:
Ok. I think I've managed to figure out a way to show the dialog only when appropriate.

Try this:

```suggestion
        if (curLatLng == null) {
            if (!locationManager.isLocationEnabled()) {
                // Location is not set to High Accuracy.
                // Let's prompt the user using a dialog box.
                showEnableLocationDialog();
            }

            return;
        }
```
Let me know if this works fine.

Reply:
Ignore my previous comment. I think a more appropriate place to make the change in would be the `NearbyParentFragmentPresenter`. We shouldn't have any concerns about handling the `null` case if we make the change there. To be clear, I'm not suggesting to do the change there to avoid the `null` handling but it really does seem to be appropriate place to do the change. Changing it here has some other unintended consequences.

I'll let you figure out the appropriate place to change there, though. 😉

Reply:
> I think a more appropriate place to make the change in would be the `NearbyParentFragmentPresenter`.

Wait a second. Don't we need to check the location state when the `Locate Me` button is pressed? And since we are using the MVP architecture, the only way to do that is in the  `View` section and `NearbyParentFragment` is the view.
We can't do this in the presenter if we want to stick to MVP architecture and even if we do it, it will be more complex than doing that in `View`.
These are just my thoughts according to what I know. That being said, like always, I could be wrong.

Reply:
I can't speak to this implementation much but if the View has logic in it then it is not a View. A View tells the Presenter something happened and then it is the Presenter's responsibility to figure out what the view should do, this may entail some updates to the Model

Reply:
But when the user presses Locate Me button, the method `ReCenterMap` checks screen orientation and also checks if the location is not null. All this is done in `NearbyParentFragment` and not in its presenter `NearbyParentFragmentPresenter`. Am  I missing something?

Reply:
Also, can you tell me which class will be called the View of the Presenter `NearbyParentFragmentPresenter`? Is it `NearbyParentFragment` or `NearbyParentFragmentContract` or any other class? I haven't worked with the MVP architecture before, so I apologize for the trouble.

Reply:
A bad design decision, they happen for many reasons and sometimes even good ones. The presenter should only call recenter if the location was not null, it should also probably construct those CameraPosition objects.

Calling things Views/Presenters/Model doesn't make them actually adhere to the rules.

Nowadays I prefer MVI because of explicit predictable state modelling but that is neither here nor there.

Reply:
And the presenter should not be setting clickListeners, the primary things you do in a view is render and set clickListeners, that should be it. The dumber the view the happier your life

Reply:
> A bad design decision, they happen for many reasons and sometimes even good ones.

I agree. Sometimes, rules have to be broken as no rule/theory/architecture can satisfy all conditions.

>The presenter should only call recenter if the location was not null, it should also probably construct those CameraPosition objects.

Maybe we can open a new issue for it.

>Nowadays I prefer MVI because of explicit predictable state modelling but that is neither here nor there.

Thanks, I will check MVI out. Intents seem great.

Reply:
> The dumber the view the happier your life

I will keep that in mind.

Reply:
[This article](https://medium.com/@cervonefrancesco/model-view-presenter-android-guidelines-94970b430ddf) isn't bad. I haven't done MVP in years so I don't have immediate access to a great resource.

Reply:
> [This article](https://medium.com/@cervonefrancesco/model-view-presenter-android-guidelines-94970b430ddf) isn't bad. I haven't done MVP in years so I don't have immediate access to a great resource.

Thanks, I will check it out :)

Reply:
Sorry, your comments weren't showing up for me .

>Also, can you tell me which class will be called the View of the Presenter NearbyParentFragmentPresenter? Is it NearbyParentFragment or NearbyParentFragmentContract or any other class?

`*Contract` are typically where people put all the methods that the View/Presenter/Model will call. These `contracts` are interfaces that should describe easily what a screen does. The concrete instances are the implementations of your M/V/P, I think it is `ctrl+alt+b` to go to implementations of an interface in AS. Here it is `NearbyParentFragment` because it implements `Contract.View`

> Maybe we can open a new issue for it.

I wouldn't worry about it for now, sometimes it is better to let sleeping dogs lie. Hopefully at some point it can be addressed but I doubt it is a priority

Reply:
> I wouldn't worry about it for now, sometimes it is better to let sleeping dogs  lie. Hopefully at some point it can be addressed but I doubt it is a priority

Good point but wouldn't be nice if we let people know that there's a dog lying here? 😉

By which I mean wouldn't it be nice if we open an issue just to note that there's this issue with the way things are right now and we would have to fix them some time in the future... *before the dog wakes up*.

Reply:
NearbyParentFragment is 1456 lines long, there are other problems that are way more major than  "This method should adhere more to the architecture". By all means create the issue, I just have never seen these issues get resolved piecemeal